iron-focusables-helper.html

```diff
@@ -0,0 +1,135 @@
+<!doctype html>
+<!--
+@license
+Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+<link rel="import" href="../polymer/polymer.html">
+
+<script>
+  (function() {
+    'use strict';
+
+    Polymer.IronFocusablesHelper = {
+
+      /**
+       * Returns a sorted array of tabbable nodes, including the root node.
+       * It searches the tabbable nodes in the light and shadow dom of the chidren,
+       * sorting the result by tabindex.
+       * @param {!Node} node
+       * @return {Array<Node>}
+       */
+      getTabbableNodes: function(node) {
+        var result = [];
+        // If there is at least one element with tabindex > 0, we need to sort
+        // the final array by tabindex.
+        var maxTabIndex = this._findTabbableNodes(node, result);
+        maxTabIndex > 0 && result.sort(this._tabIndexSort);
+        return result;
+      },
+
+      /**
+       * Returns if a node is focusable.
+       * @param {!Node} node
+       * @return {boolean}
+       */
+      isFocusable: function(node) {
+        // shadow roots won't have this method.
+        if (!node.hasAttribute) {
+          return false;
+        }
+        var name = node.localName;
+        // Elements that can be focused even if they have [disabled] attribute.
+        if (name === 'iframe' ||
+          ((name === 'a' || name === 'area') && node.hasAttribute('href')) ||
+          node.hasAttribute('tabindex') || node.hasAttribute('contentEditable')) {
+          return true;
+        }
+        // Elements that cannot be focused if they have [disabled] attribute.
+        return /^(input|select|textarea|button|object)$/.test(name) && !node.disabled;
+      },
+
+      /**
+       * Returns if a node is tabbable.
+       * @param {!Node} node
+       * @return {boolean}
+       */
+      isTabbable: function(node) {
+        return node.tabIndex >= 0 && this.isFocusable(node) && this._isVisible(node);
+      },
+
+      /**
+       * Searches for nodes that are tabbable and adds them to the `result` array.
+       * Returns the maximum tabindex of the added nodes.
+       * @param {!Node} node The starting point for the search; it's added if tabbable.
+       * @param {!Array<Node>} result
+       * @return {Number}
+       * @private
+       */
+      _findTabbableNodes: function(node, result) {
```

Comment:
Just to recap:
- we don't want to sort if all elements have `tabindex = 0` -> look if there is one element with `tabIndex > 0`
- we have to traverse all the children to find the tabbables -> minimum is `O(n)`
- to filter the tabbable elements, we must look at their `tabindex` to be `>= 0`

Let's compare the two options in some scenarios for `n` nodes of which `m` are tabbable, and `i` is the index of the tabbable with `tabindex > 0`:
```javascript
// option 1
var tabbables = [];
var maxTabIndex = this._findTabbables(node, tabbables);

// option 2
var tabbables = this._findTabbables(node);
var maxTabIndex = this._findMaxTabIndex(tabbables);
```
#### scenario 1 (worst case scenario)
 `m = n` (all are tabbable), `i = -1` (no tabbable with `tabindex > 0`)
- option 1: `O(n)`
- option 2: `O(n) + O(n)` (needs to search until the end of tabbables)

#### scenario 2
`m = n/10` (only 10% tabbable), `i = -1` (no tabbable with `tabindex > 0`)
- option 1: `O(n)`
- option 2: `O(n) + O(n/10)` (needs to search until the end of tabbables)

#### scenario 3
`m = n/10` (only 10% tabbable), `i = m/2` (the tabbable on the middle has `tabindex > 0`)
- option 1: `O(n)`
- option 2: `O(n) + O(n/20)` (can stop as soon as it finds the tabbable with `tabindex > 0`)

#### scenario 4 (best case scenario)
`m = 0` (no tabbables) 
- option 1: `O(n)`
- option 2: `O(n)`